# HG changeset patch
# User Roland Haas <roland.haas@physics.gatech.edu>
# Date 1315524948 14400
# Node ID d87fce06a3cde72fae58b401604e6d4e1e007cdd
# Parent  a753e4844d1eaa1d4a895fac0246f8f33acfcf46
Carpet: Add backtrace output to file on fatal signals
Modified version of Ian Hinder's and Justin Luitjens' code.

When a Cactus simulation aborts with a signal, it is often difficult to
determine which part of the code led to the problem. The attached patch
registers a signal handler on Carpet startup for signals 11 and 6
(segmentation fault and abort, e.g. from assert()) which outputs a stack
backtrace from each process to a file, including demangling symbol names. It
uses some low-level and possibly unofficial APIs, and is likely not completely
portable. However, I have tested it on Mac OS (gcc) and Linux (intel) and it
works in those places.

Part of this code was contributed by Justin Luitjens at the Carpet developers'
workshop in summer 2010.

diff -r a753e4844d1e -r d87fce06a3cd Carpet/Carpet/src/Initialise.cc
--- a/Carpet/Carpet/src/Initialise.cc	Thu Sep 01 12:30:07 2011 -0500
+++ b/Carpet/Carpet/src/Initialise.cc	Thu Sep 08 19:35:48 2011 -0400
@@ -1,9 +1,14 @@
+#define _GNU_SOURCE 1 // needed for dladdr, best at the top to avoid inconsistent includes
+
 #include <cassert>
 #include <cstdlib>
+#include <cstring>
 #include <iostream>
 #include <map>
 #include <string>
 #include <sstream>
+#include <signal.h>
+#include <fstream>
 
 #include <cctk.h>
 #include <cctk_Parameters.h>
@@ -11,6 +16,21 @@
 #include <cctki_ScheduleBindings.h>
 #include <cctki_WarnLevel.h>
 
+// These are used for backtraces. HAVE_XXX requires cctk.h
+#include <sys/types.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#ifdef HAVE_EXECINFO_H
+#include <execinfo.h>
+#endif
+#ifdef HAVE_DLADDR
+#include <dlfcn.h>
+#endif
+#ifdef HAVE___CXA_DEMANGLE
+#include <cxxabi.h>
+#endif
+
 #include <carpet.hh>
 #include <Timers.hh>
 #include <TimerSet.hh>
@@ -41,6 +61,9 @@
   static void Initialise3tl (cGH * cctkGH);
   
   static void print_internal_data ();
+  static void signal_handler(int signum);
+  static void write_backtrace_file(void);
+  static void generate_backtrace(ostream &stacktrace);
   
   static void ScheduleTraverse
   (char const * where, char const * name, cGH * cctkGH);
@@ -56,7 +79,10 @@
     int const convlev = 0;
     cGH * const cctkGH = CCTK_SetupGH (fc, convlev);
     CCTKi_AddGH (fc, convlev, cctkGH);
-    
+
+    signal(6, signal_handler);
+    signal(11, signal_handler);
+
     do_global_mode = true;
     do_early_global_mode = true;
     do_late_global_mode = true;
@@ -1201,6 +1227,88 @@
   //////////////////////////////////////////////////////////////////////////////
   //////////////////////////////////////////////////////////////////////////////
   
+  /// Generate a stack backtrace and send it to the given output
+  /// stream
+  void generate_backtrace(ostream &stacktrace)
+  {
+    const int MAXSTACK = 100;
+    static void *addresses[MAXSTACK];
+
+    stacktrace << "Backtrace from rank " << dist::rank() << " pid " << getpid() << ":" << endl;
+
+    int n = 0;
+#ifdef HAVE_BACKTRACE
+    n = backtrace(addresses, MAXSTACK);
+#endif
+    if (n < 2) {
+      stacktrace << "Backtrace not available!\n";
+    } else {
+      stacktrace.flags(ios::hex);
+#ifdef HAVE_BACKTRACE_SYMBOLS
+      char **names = backtrace_symbols( addresses, n );
+      for (int i = 2; i < n; i++) {
+#ifdef HAVE_DLADDR
+        Dl_info info;
+#endif
+        char *demangled = NULL;
+
+        // Attempt to demangle this if possible
+        // Get the nearest symbol to feed to demangler
+#ifdef HAVE_DLADDR
+        if(dladdr(addresses[i], &info) != 0) {
+          int stat;
+          // __cxa_demangle is a naughty obscure backend and no
+          // self-respecting person would ever call it directly. ;-)
+          // However it is a convenient glibc way to demangle syms.
+#ifdef HAVE___CXA_DEMANGLE
+          demangled = abi::__cxa_demangle(info.dli_sname,0,0,&stat);
+#endif
+        }
+#endif
+
+        if (demangled != NULL) {
+
+          // Chop off the garbage from the raw symbol
+          char *loc = strchr(names[i], '(');
+          if (loc != NULL) *loc = '\0';
+
+          stacktrace << i - 1 << ". " << demangled << "(" << names[i] << ")" << '\n';
+          free(demangled);
+        } else { // Just output the raw symbol
+          stacktrace << i - 1 << ". " << names[i] << '\n';
+        }
+      }
+      free(names);
+#endif
+    }
+  }
+
+  /// Output a stack backtrace file backtrace.<rank>.txt
+  void write_backtrace_file(void)
+  {
+    // declaring parameters is "safe" since it really only accesses some hidden
+    // global structures, and no fragile linked list or similar
+    DECLARE_CCTK_PARAMETERS;
+
+    ofstream myfile;
+    stringstream ss;
+
+    ss << out_dir << "/" << "backtrace." << dist::rank() << ".txt";
+    string filename = ss.str();
+
+    cerr << "Writing backtrace to " << filename << endl;
+    myfile.open(filename.c_str());
+    generate_backtrace(myfile);
+    myfile.close();
+  }
+
+  void signal_handler(int signum)
+  {
+    cerr << "Rank " << dist::rank() << " with PID " << getpid() << " got signal" << signum << endl;
+    signal(signum, SIG_DFL); // Restore the default signal handler
+    write_backtrace_file();
+    kill(getpid(), signum); // Re-raise the signal to be caught by the default handler
+  }
   
   
   void
